### PR TITLE
SQS Secure Transport

### PR DESCRIPTION
**Pull request by kylehuntsman**
Ticket: [RUI-1379](https://jira.meltdev.com/browse/RUI-1379)

Added the ability to send sensitive data through SQS via encryption. Sensitive data is handed to the SQS `sendMessage()` function, and it will encrypt it, add it to the message body, and send it to the queue. On the consumer side, it will take the message, decrypt it, and send it along like normal. The user should not have to worry about writing encryption code, other than setting up the encryption configuration values.

### Overview
- The original SQS `sendMessage()` function now takes an optional second argument for data that should be encrypted.
- Encrypted data is attached to the original message body as a child object on the `encrypted` key. This child object has a data property, which is the encrypted data, and a key property, which is an encrypted version of the key used to encrypt the data.
- Encryption is done via AWS KMS.
- A new `encryption` property was added to the config to allow customization of the KMS key name to use for encryption. The configuration values must be set for the data to be encrypted and sent along.
  - key: The name of the KMS key to generate a key from, which will be used for the encryption.
- If data is passed without the configuration set, an `EncryptionConfigurationError` is thrown.

```js
// Configuration
{
  awsext: {
    sqs: {
      my-consumer: {
        queue: { ... },
        sqsOptions: { ... },
        kmsOptions: { ... }, // <-- Options to be passed to the KMS instance
        encryption: { // <-- Encryption configuration for sending sensitive data through SQS
          key: 'Key for SQS secure transport' // <-- KMS master key name
        }
      }
    }
  }
}
```
```js
// SQS Message Body with Encrypted Data
{
  myProperty: 'myValue',
  encrypted: {
    data: 'Ab48EF303fbeAcd908tKj2erMnd3',
    key: 'Bc59FG414gcdBde019uLk3dsNoe4'
  }
}
```

### CODE REVIEW

File: lib/sqs/sqs-base.js [hunk 1]
```diff
@@ -89,19 +93,40 @@ class SqsBase extends events.EventEmitter {
...
+              .then(() => this._encrypt(sensitiveMsgData)) // encrypt the message...
+              .then(encryptedPayload => _.merge({}, msgData, {encrypted: encryptedPayload})); // and attach it to the message
+          } else {
+            throw new error.EncryptionConfigurationError(this._encryption);
+          }
+        }
+
+        // else just validate the normal message
+        return this.validateMessage(msgData)
+          .then(() => msgData);
```
Comment (sukrit007) [inline lib/sqs/sqs-base.js:113]: Not sure why we need this line

File: lib/sqs/sqs-base.js [hunk 2]
```diff
@@ -5,11 +5,13 @@ require('promise.prototype.finally');
 const
   promisify = require("es6-promisify"),
   config = require("config"),
+  crypto = require('crypto'),
```
Comment (sukrit007) [inline lib/sqs/sqs-base.js]: Not needed

File: test/unit/common/encryption-spec.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    describe('the return object', () => {
+      it('has the correct properties', () => {
+        return promise.then(payload => {
+          payload.should.have.all.keys(['data', 'key']);
+        });
+      });
+
+      it('has the correct encrypted data', () => {
+        return promise.then(payload => {
+          payload.data.should.eql(ENCRYPTED_DATA);
```
Comment (sukrit007) [inline test/unit/common/encryption-spec.js]: We should not compare encrypted data and possibly use a random salt for encryption

File: lib/sqs/sqs-base.js (hunk 1 above)
Comment (sukrit007) [reply lib/sqs/sqs-base.js:113]: My bad .  It is clear now. THanks

File: examples/config/default.yaml [hunk 4]
```diff
@@ -4,13 +4,18 @@ awsext:
...
         prefix: 'sqs-consumer-example-'
       sqsOptions:
-        region: us-west-2
+        region: 'us-west-2'
         # endpoint: http://localhost:9324
+      kmsOptions:
+        region: 'us-east-1'
+        # endpoint: http://localhost:9324
+      encryption:
+        key: 'alias/picard-sqs'
```
Comment (sukrit007) [inline examples/config/default.yaml]: can you rename picard by example-sqs

File: lib/common/encryption.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        let key = decryptedKey.Plaintext.toString('hex');
+        decryptedKey.Plaintext = null; // We want to explicitly wipe the plain text key from memory
+
+        let bytes  = CryptoJS.AES.decrypt(encryptedPayload.data, key);
+        let dataString = bytes.toString(CryptoJS.enc.Utf8);
+
+        try {
+          return JSON.parse(dataString);
+        } catch(e) {
+          return dataString;
```
Comment (sukrit007) [inline lib/common/encryption.js:82]: @FunnyGopher  any reason why we are returning data string and not throwing exception?

**Review by sukrit007 — APPROVED** [review]
LGTM!

File: lib/common/encryption.js (hunk 5 above)
Comment (kylehuntsman) [reply lib/common/encryption.js:82]: This is incase the data being decrypted was instead a string instead of a json object

**Pull request merged**